File: set.c
#include "copyright.h"

/* commands which set parameters */
#include <stdio.h>
#include <ctype.h>

#include "db.h"
#include "config.h"
#include "match.h"
#include "interface.h"
#include "externs.h"

static dbref match_controlled(dbref player, const char *name)
{
    dbref match;

    init_match(player, name, NOTYPE);
    match_everything();

    match = noisy_match_result();
    if(match != NOTHING && !controls(player, match)) {
	notify(player, "Permission denied.");
	return NOTHING;
    } else {
	return match;
    }
}

void do_name(dbref player, const char *name, char *newname)
{
    dbref thing;
    void free(void *);
    char *password;

    if((thing = match_controlled(player, name)) != NOTHING) {
	/* check for bad name */
	if(*newname == '\0') {
	    notify(player, "Give it what new name?");
	    return;
	}

	/* check for renaming a player */
	if(Typeof(thing) == TYPE_PLAYER) {
	    /* split off password */
	    for(password = newname; isgraph(*password); password++);

	    /* eat whitespace */
	    if(*password) {
		*password++ = '\0'; /* terminate name */
		while(*password && isspace(*password)) password++;
	    }

	    /* check for null password */
	    if(!*password) {
		notify(player,
		       "You must specify a password to change a player name.");
		notify(player, "E.g.: name player = newname password");
		return;
	    } else if(strcmp(password, db[thing].password)) {
		notify(player, "Incorrect password.");
		return;
	    } else if(!payfor(player, LOOKUP_COST)
		      || !ok_player_name(newname)) {
		notify(player, "You can't give a player that name.");
		return;
	    }
	} else {
	    if(!ok_name(newname)) {
		notify(player, "That is not a reasonable name.");
		return;
	    }
	}

	/* everything ok, change the name */
	if(db[thing].name) {
	    free((void *) db[thing].name);
	}
	db[thing].name = alloc_string(newname);
	notify(player, "Name set.");
    }
}

void do_describe(dbref player, const char *name, const char *description)
{
    dbref thing;

    if((thing = match_controlled(player, name)) != NOTHING) {
	if(db[thing].description) {
	    free((void *) db[thing].description);
	}
	db[thing].description = alloc_string(description);
	notify(player, "Description set.");
    }
}

void do_fail(dbref player, const char *name, const char *message)
{
    dbref thing;

    if((thing = match_controlled(player, name)) != NOTHING) {
	if(db[thing].fail_message) {
	    free((void *) db[thing].fail_message);
	}
	db[thing].fail_message = alloc_string(message);
	notify(player, "Message set.");
    }
}

void do_success(dbref player, const char *name, const char *message)
{
    dbref thing;

    if((thing = match_controlled(player, name)) != NOTHING) {
	if(db[thing].succ_message) {
	    free((void *) db[thing].succ_message);
	}
	db[thing].succ_message = alloc_string(message);
	notify(player, "Message set.");
    }
}

void do_osuccess(dbref player, const char *name, const char *message)
{
    dbref thing;

    if((thing = match_controlled(player, name)) != NOTHING) {
	if(db[thing].osuccess) {
	    free((void *) db[thing].osuccess);
	}
	db[thing].osuccess = alloc_string(message);
	notify(player, "Message set.");
    }
}

void do_ofail(dbref player, const char *name, const char *message)
{
    dbref thing;

    if((thing = match_controlled(player, name)) != NOTHING) {
	if(db[thing].ofail) {
	    free((void *) db[thing].ofail);
	}
	db[thing].ofail = alloc_string(message);
	notify(player, "Message set.");
    }
}

void do_lock(dbref player, const char *name, const char *keyname)
{
    dbref thing;
    dbref key;
    int antilock;

    init_match(player, name, NOTYPE);
    match_everything();

    switch(thing = match_result()) {
      case NOTHING:
	notify(player, "I don't see what you want to lock!");
	return;
      case AMBIGUOUS:
	notify(player, "I don't know which one you want to lock!");
	return;
      default:
	if(!controls(player, thing)) {
	    notify(player, "You can't lock that!");
	    return;
	}
	break;
    }

    /* now we know it's ok to lock */
    if(antilock = (*keyname == NOT_TOKEN)) {
	/* skip past ! and any following whitespace */
	for(keyname++; *keyname && isspace(*keyname); keyname++);
    }
    
    /* match keyname */
    init_match(player, keyname, TYPE_THING);
    match_neighbor();
    match_possession();
    match_me();
    match_player();
    if(Wizard(player)) match_absolute();

    switch(key = match_result()) {
      case NOTHING:
	notify(player, "I can't find that key!");
	return;
      case AMBIGUOUS:
	notify(player, "I don't know which key you want!");
	return;
      default:
	if(Typeof(key) != TYPE_PLAYER
	   && Typeof(key) != TYPE_THING) {
	    notify(player, "Keys can only be players or things.");
	    return;
	}
	break;
    }
	
    /* everything ok, do it */
    db[thing].key = key;
    if(antilock) {
	db[thing].flags |= ANTILOCK;
	notify(player, "Anti-Locked.");
    } else {
	db[thing].flags &= ~ANTILOCK;
	notify(player, "Locked.");
    }
}

void do_unlock(dbref player, const char *name)
{
    dbref thing;

    if((thing = match_controlled(player, name)) != NOTHING) {
	db[thing].key = NOTHING;
	db[thing].flags &= ~ANTILOCK;
	notify(player, "Unlocked.");
    }
}

void do_unlink(dbref player, const char *name)
{
    dbref exit;

    init_match(player, name, TYPE_EXIT);
    match_exit();
    match_here();
    if(Wizard(player)) {
	match_absolute();
    }

    switch(exit = match_result()) {
      case NOTHING:
	notify(player, "Unlink what?");
	break;
      case AMBIGUOUS:
	notify(player, "I don't know which one you mean!");
	break;
      default:
	if(!controls(player, exit)) {
	    notify(player, "Permission denied.");
	} else {
	    switch(Typeof(exit)) {
	      case TYPE_EXIT:
		db[exit].location = NOTHING;
		notify(player, "Unlinked.");
		break;
	      case TYPE_ROOM:
		db[exit].location = NOTHING;
		notify(player, "Dropto removed.");
		break;
	      default:
		notify(player, "You can't unlink that!");
		break;
	    }
	}
    }
}

void do_chown(dbref player, const char *name, const char *newobj)
{
    dbref thing;
    dbref owner;

    if(!Wizard(player)) {
	notify(player, "Permission denied.");
    } else {
	init_match(player, name, NOTYPE);
	match_everything();

	if((thing = noisy_match_result()) == NOTHING) {
	    return;
	} else if((owner = lookup_player(newobj)) == NOTHING) {
	    notify(player, "I couldn't find that player.");
	} else if(Typeof(thing) == TYPE_PLAYER) {
	    notify(player, "Players always own themselves.");
	} else {
	    db[thing].owner = owner;
	    notify(player, "Owner changed.");
	}
    }
}

void do_set(dbref player, const char *name, const char *flag)
{
    dbref thing;
    const char *p;
    object_flag_type f;

    /* find thing */
    if((thing = match_controlled(player, name)) == NOTHING) return;

    /* move p past NOT_TOKEN if present */
    for(p = flag; *p && (*p == NOT_TOKEN || isspace(*p)); p++);

    /* identify flag */
    if(*p == '\0') {
	notify(player, "You must specify a flag to set.");
	return;
    } else if(string_prefix("LINK_OK", p)) {
	f = LINK_OK;
    } else if(string_prefix("DARK", p)) {
	f = DARK;
    } else if(string_prefix("STICKY", p)) {
	f = STICKY;
    } else if(string_prefix("WIZARD", p)) {
	f = WIZARD;
    } else if(string_prefix("TEMPLE", p)) {
	f = TEMPLE;
#ifdef RESTRICTED_BUILDING
    } else if(string_prefix("BUILDER", p)) {
	f = BUILDER;
#endif /* RESTRICTED_BUILDING */
    } else {
	notify(player, "I don't recognized that flag.");
	return;
    }

    /* check for restricted flag */
    if(!Wizard(player)
       && (f == WIZARD
#ifdef RESTRICTED_BUILDING
	   || f == BUILDER
#endif /* RESTRICTED_BUILDING */
	   || f == TEMPLE
	   || f == DARK && Typeof(thing) != TYPE_ROOM)) {
	notify(player, "Permission denied.");
	return;
    }

    /* check for stupid wizard */
    if(f == WIZARD && *flag == NOT_TOKEN && thing == player) {
	notify(player, "You cannot make yourself mortal.");
	return;
    }

    /* else everything is ok, do the set */
    if(*flag == NOT_TOKEN) {
	/* reset the flag */
	db[thing].flags &= ~f;
	notify(player, "Flag reset.");
    } else {
	/* set the flag */
	db[thing].flags |= f;
	notify(player, "Flag set.");
    }
}

File: lib/set.rb
require_relative 'helpers'

module MangledMud

  # Handles applying values to {Db} {Record}'s
  #
  # @version 1.0
  class Set
    include Helpers

    # @param [Db] db the current database instance
    # @param [Object] notifier An object with method do_notify(player_id, string), the method will be called to send notifications back to the player invoking the command/action
    def initialize(db, notifier)
      @db = db
      @notifier = notifier
      @match = Match.new(@db, notifier)
      @predicates = Predicates.new(@db, notifier)
      @player = Player.new(@db, notifier)
    end

    # Changes the name of something, calling back on the notifier passed into the initializer.
    #
    # @param [Number] player the database record number for the player executing the change
    # @param [String] name the name of the object to rename
    # @param [String] newname the new name to be given to the object (for players a password must also be provided)
    def do_name(player, name, newname)

      thing = match_controlled(player, name)

      if (thing != NOTHING)
        if (newname.nil? or newname.empty?) # check for bad name
          @notifier.do_notify(player, Phrasebook.lookup('what-name'))
          return
        end

        # check for renaming a player
        if (is_player(thing))
          m = newname.match(/([[:graph:]]+)\s+([[:graph:]]+)/)
          newname = (m and m[1]) ? m[1] : nil
          player_password = (m and m[2]) ? m[2] : nil

          # check for null password
          if (player_password.nil?)
            @notifier.do_notify(player, Phrasebook.lookup('specify-a-password'))
            @notifier.do_notify(player, Phrasebook.lookup('help-player-password'))
            return
          elsif (player_password != @db[thing].password)
            @notifier.do_notify(player, Phrasebook.lookup('bad-password'))
            return
          elsif (!@predicates.payfor(player, LOOKUP_COST) || !@predicates.ok_player_name(newname))
            @notifier.do_notify(player, Phrasebook.lookup('bad-player-name'))
            return
          end
        else # A thing
          if (!@predicates.ok_name(newname))
            @notifier.do_notify(player, Phrasebook.lookup('not-a-reasonable-name'))
            return
          end
        end

        # everything ok, change the name
        @db[thing].name = newname
        @notifier.do_notify(player, Phrasebook.lookup('name-set'))
      end
    end

    # Change the description of a thing, calling back on the notifier passed into the initializer.
    #
    # @param [Number] player the database record number for the player executing the change
    # @param [String] name the name of the object whose description is being changed
    # @param [String] description the new description
    # @see Record#description
    def do_describe(player, name, description)
      thing = match_controlled(player, name)
      if (thing != NOTHING)
        @db[thing].description = description
        @notifier.do_notify(player, Phrasebook.lookup('desc-set'))
      end
    end

    # Change the fail message for a thing, calling back on the notifier passed into the initializer.
    #
    # @param [Number] player the database record number for the player executing the change
    # @param [String] name the name of the object whose fail message is being changed
    # @param [String] message the new message
    # @see Record#fail
    def do_fail(player, name, message)
      thing = match_controlled(player, name)
      if (thing != NOTHING)
        @db[thing].fail = message
        @notifier.do_notify(player, Phrasebook.lookup('message-set'))
      end
    end

    # Change the success message for a thing, calling back on the notifier passed into the initializer.
    #
    # @param [Number] player the database record number for the player executing the change
    # @param [String] name the name of the object whose success message is being changed
    # @param [String] message the new message
    # @see Record#succ
    def do_success(player, name, message)
      thing = match_controlled(player, name)
      if (thing != NOTHING)
        @db[thing].succ = message
        @notifier.do_notify(player, Phrasebook.lookup('message-set'))
      end
    end

    # Change the external success message for a thing, calling back on the notifier passed into the initializer.
    #
    # @param [Number] player the database record number for the player executing the change
    # @param [String] name the name of the object whose external success message is being changed
    # @param [String] message the new message
    # @see Record#osucc
    def do_osuccess(player, name, message)
      thing = match_controlled(player, name)
      if (thing != NOTHING)
        @db[thing].osucc = message
        @notifier.do_notify(player, Phrasebook.lookup('message-set'))
      end
    end

    # Change the external fail message for a thing, calling back on the notifier passed into the initializer.
    #
    # @param [Number] player the database record number for the player executing the change
    # @param [String] name the name of the object whose external fail message is being changed
    # @param [String] message the new message
    # @see Record#ofail
    def do_ofail(player, name, message)
      thing = match_controlled(player, name)
      if (thing != NOTHING)
        @db[thing].ofail = message
        @notifier.do_notify(player, Phrasebook.lookup('message-set'))
      end
    end

    # Lock a thing to a given key, calling back on the notifier passed into the initializer.
    #
    # @param [Number] player the database record number for the player executing the change
    # @param [String] name the name of the object to lock
    # @param [String] keyname the key to lock against, example key "!*murf"
    # @see Record#key
    # @see Record#flags
    def do_lock(player, name, keyname)
      @match.init_match(player, name, NOTYPE)
      @match.match_everything()

      thing = @match.match_result()
      case thing
      when NOTHING
        @notifier.do_notify(player, Phrasebook.lookup('dont-see-lock'))
        return
      when AMBIGUOUS
        @notifier.do_notify(player, Phrasebook.lookup('which-one-lock'))
        return
      else
        if (!@predicates.controls(player, thing))
          @notifier.do_notify(player, Phrasebook.lookup('bad-lock'))
          return
        end
      end

      # now we know it's ok to lock
      if keyname
        antilock = (keyname[0] == NOT_TOKEN)
        if (antilock)
          keyname = keyname[1..-1]
          keyname.lstrip!() if keyname
        end
      end

      # match keyname
      @match.init_match(player, keyname, TYPE_THING)
      @match.match_neighbor()
      @match.match_possession()
      @match.match_me()
      @match.match_player()
      @match.match_absolute if (is_wizard(player))

      key = @match.match_result()
      case key
      when NOTHING
        @notifier.do_notify(player, Phrasebook.lookup('no-key'))
        return
      when AMBIGUOUS
        @notifier.do_notify(player, Phrasebook.lookup('which-key'))
        return
      else
        if (!is_player(key) && !is_thing(key))
          @notifier.do_notify(player, Phrasebook.lookup('bad-key-link'))
          return
        end
      end

      # everything ok, do it
      @db[thing].key = key
      if (antilock)
        @db[thing].flags |= ANTILOCK
        @notifier.do_notify(player, Phrasebook.lookup('anti-locked'))
      else
        @db[thing].flags &= ~ANTILOCK
        @notifier.do_notify(player, Phrasebook.lookup('locked'))
      end
    end

    # Unlock a thing, calling back on the notifier passed into the initializer.
    #
    # @param [Number] player the database record number for the player executing the change
    # @param [String] name the name of the object to unlock
    # @see Record#key
    # @see Record#flags
    def do_unlock(player, name)
      thing = match_controlled(player, name)
      if (thing != NOTHING)
        @db[thing].key = NOTHING
        @db[thing].flags &= ~ANTILOCK
        @notifier.do_notify(player, Phrasebook.lookup('unlocked'))
      end
    end

    # Unlink a thing, calling back on the notifier passed into the initializer.
    # This only effects exits and rooms
    #
    # @param [Number] player the database record number for the player executing the change
    # @param [String] name the name of the object to unlink
    # @see Record#location
    def do_unlink(player, name)
      @match.init_match(player, name, TYPE_EXIT)
      @match.match_exit()
      @match.match_here()
      @match.match_absolute() if(is_wizard(player))

      exit = @match.match_result()
      case exit
      when NOTHING
        @notifier.do_notify(player, Phrasebook.lookup('unlink-what'))
      when AMBIGUOUS
        @notifier.do_notify(player, Phrasebook.lookup('which-one'))
      else
        if (!@predicates.controls(player, exit))
          @notifier.do_notify(player, Phrasebook.lookup('no-permission'))
        else
          case typeof(exit)
          when TYPE_EXIT
            @db[exit].location = NOTHING
            @notifier.do_notify(player, Phrasebook.lookup('unlinked'))
          when TYPE_ROOM
            @db[exit].location = NOTHING
            @notifier.do_notify(player, Phrasebook.lookup('drop-to-removed'))
          else
            @notifier.do_notify(player, Phrasebook.lookup('cant-unlink-that'))
          end
        end
      end
    end

    # Change ownership of a thing, calling back on the notifier passed into the initializer.
    #
    # @param [Number] player the database record number for the player executing the change
    # @param [String] name the name of the object whose ownership is being changed
    # @param [String] newobj the new owner's name (a player)
    # @see Record#owner
    def do_chown(player, name, newobj)
      if (!is_wizard(player))
        @notifier.do_notify(player, Phrasebook.lookup('no-permission'))
      else
        @match.init_match(player, name, NOTYPE)
        @match.match_everything()
        thing = @match.noisy_match_result()
        owner = @player.lookup_player(newobj)
        if (thing == NOTHING)
          return
        elsif (owner == NOTHING)
          @notifier.do_notify(player, Phrasebook.lookup('no-player'))
        elsif (player?(thing))
          @notifier.do_notify(player, Phrasebook.lookup('own-self'))
        else
          @db[thing].owner = owner
          @notifier.do_notify(player, Phrasebook.lookup('owner-changed'))
        end
      end
    end

    # Change a flag on thing, calling back on the notifier passed into the initializer.
    #
    # @param [Number] player the database record number for the player executing the change
    # @param [String] name the name of the object whose flags are being changed
    # @param [String] flag the name of the flag to change, see {DARK}, {LINK_OK}, {STICKY}, {WIZARD}, {TEMPLE}
    # @see Record#flags
    def do_set(player, name, flag)
      # find thing
      thing = match_controlled(player, name)
      return if (thing == NOTHING)

      # move p past NOT_TOKEN if present
      unless flag.nil?
        p = flag.strip
        p = p[1..-1] if p[0] == NOT_TOKEN
      end

      # identify flag
      f = nil
      if (p.nil? or p.empty?)
        @notifier.do_notify(player, Phrasebook.lookup('specify-a-flag'))
        return
      elsif (p.casecmp("LINK_OK") == 0)
        f = LINK_OK
      elsif (p.casecmp("DARK") == 0)
        f = DARK
      elsif (p.casecmp("STICKY") == 0)
        f = STICKY
      elsif (p.casecmp("WIZARD") == 0)
        f = WIZARD
      elsif (p.casecmp("TEMPLE") == 0)
        f = TEMPLE
      else
        @notifier.do_notify(player, Phrasebook.lookup('unknown-flag'))
        return
      end

      # check for restricted flag
      if (!is_wizard(player) && (f == WIZARD || f == TEMPLE || (f == DARK && !room?(thing))))
        @notifier.do_notify(player, Phrasebook.lookup('no-permission'))
        return
      end

      # check for stupid wizard
      if (f == WIZARD && flag[0] == NOT_TOKEN && thing == player)
        @notifier.do_notify(player, Phrasebook.lookup('cant-be-mortal'))
        return
      end

      # else everything is ok, do the set
      if (flag[0] == NOT_TOKEN)
        # reset the flag
        @db[thing].flags &= ~f
        @notifier.do_notify(player, Phrasebook.lookup('flag-reset'))
      else
        # set the flag
        @db[thing].flags |= f
        @notifier.do_notify(player, Phrasebook.lookup('flag-set'))
      end
    end

    private

    # Check to see if player controls (owns) thing
    def match_controlled(player, name)
      @match.init_match(player, name, NOTYPE)
      @match.match_everything()

      match = @match.noisy_match_result()
      if (match != NOTHING && !@predicates.controls(player, match))
        @notifier.do_notify(player, Phrasebook.lookup('no-permission'))
        NOTHING
      else
        match
      end
    end

  end
end
